README.md
GraphQL Servers in OCaml
-----------------------------------------------

![Build Status](https://travis-ci.org/andreas/ocaml-graphql-server.svg?branch=master)

This repo contains a library for creating GraphQL servers in OCaml. Note that the API is still under active development.

Current feature set:

- [x] Type-safe schema design
- [x] GraphQL parser in pure OCaml using Menhir
- [x] Query execution
- [x] Introspection of schemas
- [x] Arguments for fields
- [x] Allows variables in queries
- [x] Lwt support
- [x] Async support
- [x] Example with HTTP server and GraphiQL
- [x] GraphQL Subscriptions

## Documentation

Four OPAM packages are provided:

- `graphql` provides the core functionality and is IO-agnostic. It provides a functor `Graphql.Schema.Make(IO)` to instantiate with your own IO monad.
- `graphql-lwt` provides the module `Graphql_lwt.Schema` with [Lwt](https://github.com/ocsigen/lwt) support in field resolvers.
- `graphql-async` provides the module `Graphql_async.Schema` with [Async](https://github.com/janestreet/async) support in field resolvers.
- `graphql_parser` provides query parsing functionality.
- `graphql-cohttp` allows exposing a schema over HTTP using [Cohttp](https://github.com/mirage/ocaml-cohttp).

API documentation:

- [`graphql`](https://andreas.github.io/ocaml-graphql-server/graphql)
- [`graphql-lwt`](https://andreas.github.io/ocaml-graphql-server/graphql-lwt)
- [`graphql-async`](https://andreas.github.io/ocaml-graphql-server/graphql-async)
- [`graphql_parser`](https://andreas.github.io/ocaml-graphql-server/graphql_parser)
- [`graphql-cohttp`](https://andreas.github.io/ocaml-graphql-server/graphql-cohttp)

## Examples

### GraphiQL

To run a sample GraphQL server also serving GraphiQL, do the following:

```bash
opam install dune graphql-lwt graphql-cohttp cohttp-lwt-unix
git clone git@github.com:andreas/ocaml-graphql-server.git
dune exec examples/server.exe
```

Now open [http://localhost:8080/graphql](http://localhost:8080/graphql).

### Defining a Schema

```ocaml
open Graphql

type role = User | Admin
type user = {
  id   : int;
  name : string;
  role : role;
}

let users = [
  { id = 1; name = "Alice"; role = Admin };
  { id = 2; name = "Bob"; role = User }
]

let role = Schema.(enum "role"
  ~doc:"The role of a user"
  ~values:[
    enum_value "USER" ~value:User;
    enum_value "ADMIN" ~value:Admin;
  ]
)

let user = Schema.(obj "user"
  ~doc:"A user in the system"
  ~fields:[
    field "id"
      ~doc:"Unique user identifier"
      ~typ:(non_null int)
      ~args:Arg.[]
      ~resolve:(fun info p -> p.id)
    ;
    field "name"
      ~typ:(non_null string)
      ~args:Arg.[]
      ~resolve:(fun info p -> p.name)
    ;
    field "role"
      ~typ:(non_null role)
      ~args:Arg.[]
      ~resolve:(fun info p -> p.role)
  ]
)

let schema = Schema.(schema [
  field "users"
    ~typ:(non_null (list (non_null user)))
    ~args:Arg.[]
    ~resolve:(fun info () -> users)
])
```

### Running a Query

Without variables:

```ocaml
match Graphql_parser.parse "{ users { name } }" with
| Ok query -> Graphql.Schema.execute schema ctx query
| Error err -> failwith err
```

With variables parsed from JSON:

```ocaml
match Graphql_parser.parse "{ users(limit: $x) { name } }" with
| Ok query ->
    let json_variables = Yojson.Basic.(from_string "{\"x\": 42}" |> Util.to_assoc) in
    let variables = (json_variables :> (string * Graphql_parser.const_value) list)
    Graphql.Schema.execute schema ctx ~variables query
| Error err ->
    failwith err
```

### Recursive Objects

The function `Schema.fix` can be used to define both self-recursive and mutually recursive objects:

```ocaml
(* self-recursive *)
type tweet = {
  id : int;
  replies : tweet list;
}

let tweet = Schema.(fix (fun recursive ->
  recursive.obj "tweet"
    ~fields:(fun tweet -> [
      field "id"
        ~typ:(non_null int)
        ~args:Arg.[]
        ~resolve:(fun info t -> t.id)
        ;
      field "replies"
        ~typ:(non_null (list (non_null tweet)))
        ~args:Arg.[]
        ~resolve:(fun info t -> t.replies)
    ])))
```

```ocaml
(* mutually recursive *)
let foo, bar = Schema.(fix (fun recursive ->
  let foo = recursive.obj "foo" ~fields:(fun (_, bar) -> [
      field "bar"
        ~typ:bar
        ~args:Arg.[]
        ~resolve:(fun info foo -> foo.bar)
    ])
  in
  let bar = recursive.obj "bar" ~fields:(fun (foo, _) -> [
    field "foo"
      ~typ:foo
      ~args:Arg.[]
      ~resolve:(fun info bar -> bar.foo)
      ])
  in
  foo, bar))
```

### Lwt Support

```ocaml
open Lwt.Infix
open Graphql_lwt

let schema = Schema.(schema [
  io_field "wait"
    ~typ:(non_null float)
    ~args:Arg.[
      arg "duration" ~typ:float;
    ]
    ~resolve:(fun info () ->
      Lwt_result.ok (Lwt_unix.sleep duration >|= fun () -> duration)
    )
])
```

### Async Support

```ocaml
open Core.Std
open Async.Std
open Graphql_async

let schema = Schema.(schema [
  io_field "wait"
    ~typ:(non_null float)
    ~args:Arg.[
      arg "duration" ~typ:float;
    ]
    ~resolve:(fun info () ->
      after (Time.Span.of_float duration) >>| fun () -> duration
    )
])
```

### Arguments

Arguments for a field can either be required, optional or optional with a default value:

```ocaml
Schema.(obj "math"
  ~fields:(fun _ -> [
    field "sum"
      ~typ:int
      ~args:Arg.[
        arg  "x" ~typ:(non_null int); (* <-- required *)
        arg  "y" ~typ:int;            (* <-- optional *)
        arg' "z" ~typ:int ~default:7  (* <-- optional w/ default *)
      ]
      ~resolve:(fun info () x y z ->
        let y' = match y with Some n -> n | None -> 42 in
        x + y' + z
      )
  ])
)
```

Note that you must use `arg'` to provide a default value.

### Subscriptions

```ocaml
Schema.(schema [
     ...
  ]
  ~subscriptions:[
    subscription_field "user_created"
      ~typ:(non_null user)
      ~resolve:(fun info ->
        let user_stream, push_to_user_stream = Lwt_stream.create () in
        let destroy_stream = (fun () -> push_to_user_stream None) in
        Lwt_result.return (user_stream, destroy_stream))
    ])
```

### HTTP Server

Using Lwt:

```ocaml
open Graphql_lwt

let schema = Schema.(schema [
  ...
])

module Graphql_cohttp_lwt = Graphql_cohttp.Make (Schema) (Cohttp_lwt.Body)

let () =
  let callback = Graphql_cohttp_lwt.make_callback (fun _req -> ()) schema in
  let server = Cohttp_lwt_unix.Server.make ~callback () in
  let mode = `TCP (`Port 8080) in
  Cohttp_lwt_unix.Server.create ~mode server
  |> Lwt_main.run
```

## Design

Only valid schemas should pass the type checker. If a schema compiles, the following holds:

1. The type of a field agrees with the return type of the resolve function.
2. The arguments of a field agrees with the accepted arguments of the resolve function.
3. The source of a field agrees with the type of the object to which it belongs.
4. The context argument for all resolver functions in a schema agree.

The following blog posts introduces the core design concepts:

- https://andreas.github.io/2017/11/29/type-safe-graphql-with-ocaml-part-1/
- https://andreas.github.io/2018/01/05/modeling-graphql-type-modifiers-with-gadts-part-2/
- https://andreas.github.io/2019/05/20/graphql-resolver-arguments-as-diff-lists-part-3/


CHANGES.md
0.14.0 2022-07-08
---------------------------------

- Support `__typename` on subscriptions (#178)
- Handle unknown fields for subscriptions (#178)
- Add ocamlformat (#177)
- Handle missing variables as null (#184)
- Show default value in introspection query (#194)
- Support block strings in the parser (#198)
- Handle skip/include directives on fragment spreads (#200)
- Improved handling of recursive arguments and objects (#199)
- Fix websocket conflict (#206)
- Update deprecated Fmt functions (#206)
- Use Yojson `t` types instead of deprecated `json` type (#208)
- Raise minimum `rresult` version (#209)

0.13.0 2019-09-21
---------------------------------

- Support for custom errors (#166)
- Add `__type` introspection field (#163)
- Better parameter handling for `graphql-cohttp` (#156)
- Fix use of default variables (#158)
- Fix merging of field selections (#174)

0.12.2 2019-04-02
---------------------------------

- Remove open of Result (#154)

0.12.1 2019-03-29
---------------------------------

- Disable Yojson deprecation warnings (#148)
- Fix routing in graphql-cohttp (#150)

0.12.0 2019-03-22
---------------------------------

- Remove Str from cohttp-graphql (#146)


0.11.0 2019-03-03
---------------------------------

- Minor parser improvements (#139)
- Replace Str with Re (#144)

0.10.0 2019-02-14
---------------------------------

- Fix digestif dependency in graphql-cohttp (#141)

0.9.0 2019-02-08
---------------------------------

- Skip and include directives (#117)
- Expose more data to resolvers (#127)
- Add the `errors` key first in the response JSON (#131)
- Rewrite parser to Menhir and replace sexp with fmt (#132)
- Support for websockets as transport (#133)

0.8.0 2018-12-06
---------------------------------

- Subscription support (#101)
- Add path to errors in response (#112)
- Improved escaped character handling (#114)
- Improve error messages for invalid arguments (#128)

0.7.0 2018-08-10
---------------------------------

- Allow returning errors from resolve function of io_field (#69)
- Support for union and interface types (#70)
- Expose HTTP request to context construction function in Graphql_lwt.Server.start (#88)
- Fix error response from Graphql_lwt.Server (#96)
- Allow passing operation name to Graphql_lwt.Server (#103)
- Querying undefined fields gives validation error (#105)
- Fix parsing of enums with E (#109)

0.6.0 2018-03-31
---------------------------------

- Prevent fragment cycles (#78)
- graphql-lwt depends on cohttp 1.0.0 (#71)

0.5.0 2018-01-18
---------------------------------

- Depend on angstrom 0.7.0 (#64)
- Fix parsing of quoted strings (#64)
- Make custom argument types generalizable (#72)
- Deduplicate arg types in introspection result (#75)
- 4.06 compatibility (#76)

0.4.0 2017-09-17
---------------------------------

- Parse tabs as whitespace (#62)
- Move parser to separate package (#63)

0.3.0 2017-08-31
---------------------------------

- Built-in HTTP server for graphql-lwt (#60)

0.2.0 2017-07-02
---------------------------------

- Support deprecation of fields (#53)
- Support documentation and deprecation of enum values (#54)

0.1.0 2017-05-25
---------------------------------

Initial public release.


examples/server.ml
open Lwt.Infix
open Graphql_lwt

type role = User | Admin

type user = { id : int; name : string; role : role; friends : user list }

let rec alice = { id = 1; name = "Alice"; role = Admin; friends = [ bob ] }

and bob = { id = 2; name = "Bob"; role = User; friends = [ alice ] }

let users = [ alice; bob ]

let role =
  Schema.(
    enum "role"
      ~values:
        [
          enum_value "USER" ~value:User ~doc:"A regular user";
          enum_value "ADMIN" ~value:Admin ~doc:"An admin user";
        ])

let user =
  Schema.(fix (fun recursive ->
    recursive.obj "user" ~fields:(fun user ->
        [
          field "id"
            ~args:Arg.[]
            ~typ:(non_null int)
            ~resolve:(fun _ p -> p.id);
          field "name"
            ~args:Arg.[]
            ~typ:(non_null string)
            ~resolve:(fun _ p -> p.name);
          field "role"
            ~args:Arg.[]
            ~typ:(non_null role)
            ~resolve:(fun _ p -> p.role);
          field "friends"
            ~args:Arg.[]
            ~typ:(list (non_null user))
            ~resolve:(fun _ p -> Some p.friends);
        ])))

let rec consume_stream stream =
  Lwt.catch
    (fun () ->
      Lwt_stream.next stream >>= fun x ->
      let (Ok x | Error x) = x in
      Printf.eprintf "stream response: '%s'\n%!" (Yojson.Basic.to_string x);
      if Lwt_stream.is_closed stream then Lwt.return_unit
      else consume_stream stream)
    (function
      | Lwt_stream.Closed | Lwt_stream.Empty -> Lwt.return_unit
      | _ -> Lwt.return_unit)

let set_interval s f destroy =
  let rec set_interval_loop s f n =
    let timeout =
      Lwt_timeout.create s (fun () ->
          if n > 0 then (
            f ();
            set_interval_loop s f (n - 1) )
          else destroy ())
    in
    Lwt_timeout.start timeout
  in
  set_interval_loop s f 5

let schema =
  Schema.(
    schema
      [
        io_field "users"
          ~args:Arg.[]
          ~typ:(non_null (list (non_null user)))
          ~resolve:(fun _ () -> Lwt_result.return users);
        field "greeter" ~typ:string
          ~args:
            Arg.
              [
                arg "config"
                  ~typ:
                    (non_null
                       (obj "greeter_config"
                          ~coerce:(fun greeting name -> (greeting, name))
                          ~fields:
                            [
                              arg' "greeting" ~typ:string ~default:(`String "hello");
                              arg "name" ~typ:(non_null string);
                            ]));
              ]
          ~resolve:(fun _ () (greeting, name) ->
            Some (Format.sprintf "%s, %s" greeting name));
      ]
      ~subscriptions:
        [
          subscription_field "subscribe_to_user" ~typ:(non_null user)
            ~args:Arg.[ arg' "intarg" ~typ:int ~default:(`Int 42) ]
            ~resolve:(fun _info _intarg ->
              let user_stream, push_to_user_stream = Lwt_stream.create () in
              let destroy_stream () = push_to_user_stream None in
              set_interval 2
                (fun () ->
                  let idx = Random.int (List.length users) in
                  push_to_user_stream (Some (List.nth users idx)))
                destroy_stream;
              Lwt_result.return (user_stream, destroy_stream));
        ])

module Graphql_cohttp_lwt =
  Graphql_cohttp.Make (Graphql_lwt.Schema) (Cohttp_lwt_unix.IO)
    (Cohttp_lwt.Body)

let () =
  let open Printf in
  let on_exn = function
    | Unix.Unix_error (error, func, arg) ->
        printf "Client connection error %s: %s(%S)" (Unix.error_message error)
          func arg
    | exn -> printf "Unhandled exception: %s\n%!" (Printexc.to_string exn)
  in
  let callback = Graphql_cohttp_lwt.make_callback (fun _req -> ()) schema in
  let server = Cohttp_lwt_unix.Server.make_response_action ~callback () in
  let port = 8080 in
  let mode = `TCP (`Port port) in
  printf "listening on http://localhost:%d/graphql\n%!" port;
  Cohttp_lwt_unix.Server.create ~on_exn ~mode server |> Lwt_main.run
